src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.macros
+
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.childrenOfType
+
+class MacroExpander(project: Project) {
+    val psiFactory = RsPsiFactory(project)
+
+    fun expandMacro(def: RsMacroDefinition, call: RsMacroCall): List<ExpansionResult>? {
+        val case = def.macroDefinitionBodyStubbed
+            ?.macroDefinitionCaseList
+            ?.singleOrNull()
+            ?: return null
+
+        val patGroup = case.macroPattern.macroBindingGroupList.singleOrNull()
+            ?: return null
+        val patBinding = patGroup.macroBindingList.singleOrNull()
+            ?: return null
+        val name = patBinding.colon.prevSibling?.text
+        val type = patBinding.colon.nextSibling?.text
+        if (!(name != null && type != null && type == "item")) return null
+
+        val expGroup = case.macroExpansion.macroExpansionReferenceGroupList.singleOrNull()
+            ?: return null
+        val expansionText = expGroup.textBetweenParens(expGroup.lparen, expGroup.rparen)?.toString() ?: return null
+
+        val items = parseBodyAsItemList(call) ?: return null
+        val newText = items.joinToString("\n\n") { item ->
+            expansionText.replace("$" + name, item.text)
+        }
+
+        val expandedFile = psiFactory.createFile(newText)
+
+        return expandedFile.childrenOfType()
+    }
+
+
+    private val RsMacroDefinition.macroDefinitionBodyStubbed: RsMacroDefinitionBody?
+        get() {
+            val stub = stub ?: return macroDefinitionBody
+            val text = stub.macroBody ?: return null
+            return CachedValuesManager.getCachedValue(this) {
```

Comment:
Hm, we have a lot of similar code with caches.
Maybe create extension function for that?
Something like
```rust
fun <T> RsElement.getCachedValue(provider: () -> T): T {
    return CachedValuesManager.getCachedValue(this) {
        CachedValueProvider.Result.create(
            provider(),
            PsiModificationTracker.MODIFICATION_COUNT
        )
    }
}
```

Reply:
It use java class of second parameter of `getCachedValue` as an identity (key) of our cache, so we must write it manually each time we want to create a new cache ((

Reply:
Oh, I've understood
Thanks)